File: 2022/scala/src/main/scala/Day08.scala
type Trees = List[List[Int]]

object Day08 {
  def solve = {
    val input = Util.readResource("Day08/input.txt").toList
    println(part1(input))
    println(part2(input))
  }

  def part1(input : List[String]) : Int = {
    val trees = toTrees(input)
    visibleTrees(trees).size
  }

  def part2(input : List[String]) : Int = {
    val trees = toTrees(input)
    val visibleTreesList = visibleTrees(trees)
    scenicScores(visibleTreesList, trees).max
  }

  def scenicScores(visibleTreesList : List[(Int, Int)], trees : Trees) : List[Int] = {
    visibleTreesList.map(scenicScore(_, trees))
  }

  def scenicScore(tree : (Int, Int), trees : Trees) : Int = {
    val left = trees(tree._1).take(tree._2).reverse
    val right = trees(tree._1).drop(tree._2 + 1)
    val top = trees.map(_(tree._2)).take(tree._1).reverse
    val bottom = trees.map(_(tree._2)).drop(tree._1 + 1)
    val treeHeight = trees(tree._1)(tree._2)

    val leftScore = directionalScenicScore(treeHeight, left)
    val rightScore = directionalScenicScore(treeHeight, right)
    val topScore = directionalScenicScore(treeHeight, top)
    val bottomScore = directionalScenicScore(treeHeight, bottom)

    leftScore * rightScore * topScore * bottomScore
  }

  def directionalScenicScore(height : Int, treeLine : List[Int]) : Int = {
    var score = 0
    treeLine.takeWhile(treeHeight => {
      score += 1
      treeHeight < height
    })
    score
  }

  def toTrees(lines : List[String]) : Trees = {
    lines.map(_.split("").toList.map(_.toInt))
  }

  def visibleTrees(trees : Trees) : List[(Int, Int)] = {
    var idX = -1
    var idY = -1
    trees.flatMap(row => {
      idX += 1
      idY = -1
      row.flatMap(col => {
        idY +=1
        val location = (idX, idY)
        val isVisible = checkLeft(trees, location, col)
          || checkRight(trees, location, col)
          || checkTop(trees, location, col)
          || checkBottom(trees, location, col)
        if isVisible then Some(location) else None
      })
    })
  }

  def checkLeft(trees : Trees, point : (Int, Int), height : Int) : Boolean = {
    trees(point._1)
      .take(point._2)
      .forall(otherHeight => otherHeight < height)
  }

  def checkRight(trees : Trees, point : (Int, Int), height : Int) : Boolean = {
    trees(point._1)
      .drop(point._2 + 1)
      .forall(otherHeight => otherHeight < height)
  }

  def checkTop(trees : Trees, point : (Int, Int), height : Int) : Boolean = {
    trees.map(row => row(point._2))
      .take(point._1)
      .forall(otherHeight => otherHeight < height)
  }

  def checkBottom(trees : Trees, point : (Int, Int), height : Int) : Boolean = {
    trees.map(row => row(point._2))
      .drop(point._1 + 1)
      .forall(otherHeight => otherHeight < height)
  }
}

File: 2022/rust/src/bin/day08.rs
use std::fs::read_to_string;

fn main() {
    let input = read_to_string("src/inputs/08.txt").unwrap();
    part1(&input);
    part2(&input);
}

fn part1(input: &String) {
    let map = parse_input(input);
    let points = points(&map);
    let count = points
        .iter()
        .filter(|point| is_visible(&map, point))
        .count();
    println!("{}", count);
}

fn part2(input: &String) {
    let map = parse_input(input);
    let points = points(&map);
    let best_score = points
        .iter()
        .map(|point| to_score(&map, point))
        .max()
        .unwrap();
    println!("{}", best_score);
}

fn to_score(map: &Vec<Vec<u32>>, point: &(usize, usize)) -> u32 {
    let point_height = map.get(point.1).unwrap().get(point.0).unwrap();
    let left = get_left(map, point);
    let right = get_right(map, point);
    let up = get_up(map, point);
    let down = get_down(map, point);
    let mut left_score = 0;
    for height in left.iter() {
        left_score += 1;
        if height >= point_height {
            break;
        }
    }
    let mut right_score = 0;
    for height in right.iter() {
        right_score += 1;
        if height >= point_height {
            break;
        }
    }
    let mut up_score = 0;
    for height in up.iter() {
        up_score += 1;
        if height >= point_height {
            break;
        }
    }
    let mut down_score = 0;
    for height in down.iter() {
        down_score += 1;
        if height >= point_height {
            break;
        }
    }
    left_score * right_score * up_score * down_score
}

fn is_visible(map: &Vec<Vec<u32>>, point: &(usize, usize)) -> bool {
    let point_height = map.get(point.1).unwrap().get(point.0).unwrap();
    // check left
    let left = get_left(map, point);
    if left.is_empty() || left.iter().all(|tree| tree < point_height) {
        return true;
    }
    // check right
    let right = get_right(map, point);
    if right.is_empty() || right.iter().all(|tree| tree < point_height) {
        return true;
    }
    // check up
    let up = get_up(map, point);
    if up.is_empty() || up.iter().all(|tree| tree < point_height) {
        return true;
    }
    // check down
    let down = get_down(map, point);
    if down.is_empty() || down.iter().all(|tree| tree < point_height) {
        return true;
    }

    return false;
}

fn get_left<'a>(map: &'a Vec<Vec<u32>>, point: &(usize, usize)) -> Vec<u32> {
    let row = &map[point.1];
    let mut temp = row[0..point.0].to_vec();
    temp.reverse();
    temp
}

fn get_right<'a>(map: &'a Vec<Vec<u32>>, point: &(usize, usize)) -> Vec<u32> {
    let row = &map[point.1];
    row[point.0 + 1..].to_vec()
}

fn get_up<'a>(map: &'a Vec<Vec<u32>>, point: &(usize, usize)) -> Vec<u32> {
    let col: Vec<u32> = map.iter().map(|row| row[point.0]).collect();
    let mut temp = col[0..point.1].to_vec();
    temp.reverse();
    temp
}

fn get_down<'a>(map: &'a Vec<Vec<u32>>, point: &(usize, usize)) -> Vec<u32> {
    let col: Vec<u32> = map.iter().map(|row| row[point.0]).collect();
    col[point.1 + 1..].to_vec()
}

fn points(map: &Vec<Vec<u32>>) -> Vec<(usize, usize)> {
    let mut points: Vec<(usize, usize)> = vec![];
    for (x, row) in map.iter().enumerate() {
        for (y, _) in row.iter().enumerate() {
            points.push((x, y));
        }
    }
    return points;
}

fn parse_input(input: &String) -> Vec<Vec<u32>> {
    input
        .lines()
        .map(|line| {
            line.chars()
                .map(|char| char.to_digit(10).unwrap())
                .collect::<Vec<u32>>()
        })
        .collect()
}
